### PR TITLE
Feature: Allow client and server to negotiate on compression to use for the savegame

### PR DESCRIPTION
**Pull request by ldpl**
## Motivation / Problem

Client and server can be compiled with different sets of compression libraries. If client doesn't have the compression server uses it just fails. Also game uses lzma:2 for everything which isn't the best option for the network.

## Description

Client on join sends bitset of all supported compressions to the server and lets the server to chose whichever it likes more. For that server has network.savegame_formats setting that allows specifying multiple compressions (space-separated) that will be tried in that order (if not specified it uses "zlib:2 lzma:0 lzo:0"). This also allows the server to use different compression methods for network connections and local saves. 
In a way this is also a preparation for zstd inclusion (#8773) as otherwise it will have to use the older zstd version to support old platforms.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/table/settings.ini [hunk 1]
```diff
@@ -3705,6 +3705,14 @@ min      = 1
 max      = 65535
 cat      = SC_EXPERT
 
+[SDTC_STR]
+var      = network.savegame_formats
+type     = SLE_STRB
+flags    = SLF_NOT_IN_SAVE | SLF_NO_NETWORK_SYNC
+def      = nullptr
+proc     = UpdateRconPassword
```
Comment (JGRennison) [inline src/table/settings.ini]: This proc callback is probably not necessary.

**Comment by JGRennison** [comment]
More generally, is it really necessary that the server operator is able to manually specify the algorithm order and levels?
Simply using the "correct" values for the current circumstances instead of allowing the user to specify an alternative set would probably make this simpler.
I can't imagine that the existing compression configuration field sees much, if any, use outside of testing.

**Comment by James103** [comment]
@JGRennison Some people still prefer to use a slightly larger but much faster savegame compression method (such as lzo instead of lzma) in order to reduce the CPU time required to save the game at the cost of some storage space.

Example: `savegame_format = lzo` can be up to 14x faster at the cost of only 1.75x the file size (as compared to `zlib:6`).

**Comment by ldpl** [comment]
@JGRennison My main reason was to allow disabling certain method if something goes wrong with it. Also wouldn't hurt for better DoS protection as it allows to disable slower or less efficient methods. And it doesn't rly make it simpler as it only saves a bit of parsing but requires some ifdefed constant or smth.

**Comment by ldpl** [comment]
Not that I'm completely happy with it but at least it looks decent now. Anyway I'm out of ideas for how to improve it so I guess it's ready for a review.

File: src/network/network_client.cpp [hunk 2]
```diff
@@ -363,6 +363,7 @@ NetworkRecvStatus ClientNetworkGameSocketHandler::SendJoin()
 	p->Send_string(_settings_client.network.client_name); // Client name
 	p->Send_uint8 (_network_join_as);     // PlayAs
 	p->Send_uint8 (NETLANG_ANY);          // Language
+	p->Send_uint8 (GetAvailableLoadFormats());  // Compressnion formats that we can decompress
```
Comment (glx22) [inline src/network/network_client.cpp]: Typo in comment

File: src/saveload/saveload.cpp [hunk 3]
```diff
@@ -2305,101 +2305,129 @@ struct LZMASaveFilter : SaveFilter {
...
-		SetDParamStr(0, s);
-		SetDParamStr(1, def->name);
-		ShowErrorMessage(STR_CONFIG_ERROR, STR_CONFIG_ERROR_INVALID_SAVEGAME_COMPRESSION_ALGORITHM, WL_CRITICAL);
+/**
+ * Return the bitset of savegame formats that this game instance can load
+ * @return bitset of available savegame formats
+ */
+uint8 GetAvailableLoadFormats()
+{
+	return 3;
```
Comment (ldpl) [inline src/saveload/saveload.cpp]: In case anyone else still cares about this PR, I just found this xD

**Review by glx22 — COMMENTED** [review]
And I just noticed I had started a review, but never submitted it.

**Comment by TrueBrain** [comment]
As mentioned in #8773, this PR only has merit if there is more than one compression being offered that are useful. As there currently aren't, let's close this up for now, and reopen when we ever need it again.

**Pull request closed without merging**